gk/main.c

```diff
@@ -1483,232 +1776,92 @@ process_pkts_front(uint16_t port_front, uint16_t port_back,
-						&fib->u.neigh,
-						&packet.flow.f.v4.dst);
-				} else {
-					eth_cache = lookup_ether_cache(
-						&fib->u.neigh6,
-						&packet.flow.f.v6.dst);
-				}
-
-				RTE_VERIFY(eth_cache != NULL);
-
-				if (adjust_pkt_len(pkt, back, 0) == NULL ||
-						pkt_copy_cached_eth_header(pkt,
-							eth_cache,
-							back->l2_len_out)) {
-					drop_packet_front(pkt, instance);
-					continue;
-				}
-
-				if (update_ip_hop_count(front, &packet,
-						num_pkts, icmp_bufs,
-						&instance->front_icmp_rs,
-						instance,
-						drop_packet_front) < 0)
-					continue;
-
-				tx_bufs[num_tx++] = pkt;
-				continue;
-			}
+	if (r != 0) {
+		ret = rte_hash_lookup_bulk(instance->ip_flow_hash_table,
+			(const void **)&flow_arr[q], r, &pos_arr[q]);
+		if (ret != 0) {
+			GK_LOG(NOTICE,
+				"failed to find multiple keys in the hash table at lcore %u\n",
+				rte_lcore_id());
+		}
+	}
 
-			case GK_DROP:
-				/* FALLTHROUGH */
-			default:
-				drop_packet_front(pkt, instance);
-				continue;
+	for (i = 0; i < num_ip_flows; i++) {
+		if (pos_arr[i] >= 0)
+			fe_arr[i] = &instance->ip_flow_entry_table[pos_arr[i]];
+		else {
+			fe_arr[i] = NULL;
+			if (flow_arr[i]->proto == RTE_ETHER_TYPE_IPV4) {
+				lpm_lookup_pos[num_lpm_lookups] = i;
+				flows[num_lpm_lookups] = flow_arr[i];
+				num_lpm_lookups++;
+			} else {
+				lpm6_lookup_pos[num_lpm6_lookups] = i;
+				flows6[num_lpm6_lookups] = flow_arr[i];
+				num_lpm6_lookups++;
 			}
 		}
+	}
 
-		/*
-		 * Some notes regarding flow rates and units:
-		 *
-		 * Flows in the GK_REQUEST state are bandwidth limited
-		 * to an overall rate relative to the link. Therefore,
-		 * the Ethernet frame overhead is counted toward the
-		 * credits used by requests. The request channel rate
-		 * is measured in megabits (base 10) per second to
-		 * match the units used by hardware specifications.
-		 *
-		 * Granted flows (in state GK_GRANTED or sometimes
-		 * GK_BPF) are allocated budgets that are intended
-		 * to reflect the max throughput of the flow, and
-		 * therefore do not include the Ethernet frame overhead.
-		 * The budgets of granted flows are measured in
-		 * kibibytes (base 2).
-		 */
-		switch (fe->state) {
-		case GK_REQUEST:
-			ret = gk_process_request(fe, &packet,
-				gk_conf->sol_conf, stats);
-			break;
+	/* The remaining flows need LPM lookups. */
+	lookup_fib_bulk(&gk_conf->lpm_tbl, flows, num_lpm_lookups, fibs);
+	lookup_fib6_bulk(&gk_conf->lpm_tbl, flows6, num_lpm6_lookups, fibs6);
 
-		case GK_GRANTED:
-			ret = gk_process_granted(fe, &packet,
-				gk_conf->sol_conf, stats);
-			break;
+	for (i = 0; i < num_lpm_lookups; i++) {
+		int fidx = lpm_lookup_pos[i];
 
-		case GK_DECLINED:
-			ret = gk_process_declined(fe, &packet,
-				gk_conf->sol_conf, stats);
-			break;
+		fe_arr[fidx] = lookup_fe_from_lpm(&pkt_arr[fidx],
+			flow_hash_val_arr[fidx], fibs[i], &num_tx, tx_bufs,
+			acl4, acl6, num_pkts, icmp_bufs, front, back,
+			instance, gk_conf, stats);
+	}
```

Comment:
`prefetch_flow_entry(fe_arr[fidx]);`

Reply:
We don't need to prefetch here because `lookup_fe_from_lpm()` already wrote the new flow entry.